File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def ispal(s):
            n=len(s)
            for i in range(n//2):
                if s[i]!=s[n-i-1]:
                    return False
            return True
        def solve(s,temp,Ans):
            if len(s)==0:
                Ans.append(temp)
                return
            for i in range(1,len(s)+1):
                if ispal(s[:i]):
                    solve(s[i:],temp+[s[:i]],Ans)
        Ans=[]
        solve(s,[],Ans)
        return Ans
                
                
                
                
                
                

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>>ans;
    vector<vector<string>> partition(string s) {
       ans.clear();
        solve(s,0,{});
        return ans;
    }
private:
    void solve(string s,int ind,vector<string>curr)
    {
        if(ind==s.size())ans.push_back(curr);
        for(int i=ind;i<s.size();i++)
        {
            if(ispal(s,ind,i))
            {
                curr.push_back(s.substr(ind,i-ind+1));
                solve(s,i+1,curr);
                curr.pop_back();
            }
        }
    }
    bool ispal(string s,int start,int end)
    {
        while(start<end)
        {
            if(s[start]!=s[end])return false;
            start++;
            end--;
        }
        return true;
    }
};